### PR TITLE
[yaml] add mysql, oracle, postgres and sql server providers

### PR DESCRIPTION
**Pull request by Polber**
PR adds standalone SchemaTransform Providers for the specific implementations of JDBC (MySQL, PostgreSQL, Oracle and SQL Server).

This was already simulated in the Beam YAML framework, but by implementing concrete providers, other SDKs can take advantage of the pre-packaged drivers (Python and Go SDKs). This also allows for specific doc generation specific to an implementation.

More detailed docs were also added.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by Polber** [comment]
R: @robertwb

**Comment by Polber** [comment]
R: @damccorm

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java [hunk 1]
```diff
@@ -172,71 +312,55 @@ public PCollectionRowTuple expand(PCollectionRowTuple input) {
   @DefaultSchema(AutoValueSchema.class)
   public abstract static class JdbcWriteSchemaTransformConfiguration implements Serializable {
 
-    @Nullable
-    public abstract String getDriverClassName();
```
Comment (damccorm) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java:176]: Note that reordering params like this makes this PR significantly harder to review - this would also probably be more helpful as a separate PR if there is a reason to do this kind of reordering; I think the change is fine (though I don't really understand the need), but doing it as part of a large PR with functional changes makes it hard to tell what is cosmetic and what is functional

Comment (damccorm) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java:176]: I think all the parameters changed here are just cosmetic, please let me know if that is wrong

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 2]
```diff
@@ -241,7 +368,7 @@ public abstract static class Builder {
 
       public abstract Builder setConnectionInitSql(List<String> value);
 
-      public abstract Builder setFetchSize(Short value);
+      public abstract Builder setFetchSize(Integer value);
```
Comment (damccorm) [inline sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java:371]: Why did we change the typing here?

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java (hunk 1 above)
Comment (damccorm) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java:176]: I also think the same is true of the validate change, again please correct me if that's wrong

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java [hunk 3]
```diff
@@ -54,33 +55,177 @@ public class JdbcWriteSchemaTransformProvider
...
+        prettyName, transformName, transformName, prefix, port, transformName, prefix, port);
+  }
+
   @Override
   protected @UnknownKeyFor @NonNull @Initialized Class<JdbcWriteSchemaTransformConfiguration>
       configurationClass() {
     return JdbcWriteSchemaTransformConfiguration.class;
   }
 
+  protected static void validateConfig(
```
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java:161]: I think I'd actually prefer we leave this unless there's a reason to change it since it is technically public

File: sdks/python/apache_beam/yaml/generate_yaml_docs.py [hunk 4]
```diff
@@ -196,12 +196,28 @@ def io_grouping_key(transform_name):
...
 
+def add_transform_links(transform, description, provider_list):
+  """
+  Convert references of Providers to urls that link to their respective pages.
+  Avoid self-linking within a Transform page.
+  """
+  for p in provider_list:
+    description = re.sub(
+        rf"(?<!type: )\b(?!{transform}\b)\b{p}\b",
+        f'<a href="#{p.lower()}">{p}</a>',
```
Comment (damccorm) [inline sdks/python/apache_beam/yaml/generate_yaml_docs.py:207]: I'm having a hard time following this regex, could you add a comment breaking down what it is doing please?

Comment (damccorm) [reply sdks/python/apache_beam/yaml/generate_yaml_docs.py:207]: Or maybe an example would help

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java [hunk 5]
```diff
@@ -76,40 +76,34 @@ public String description() {
...
-        + "        - type: ReadFromSpanner\n"
-        + "          config:\n"
-        + "            instance_id: 'my-instance-id'\n"
-        + "            database_id: 'my-database'\n"
-        + "            query: 'SELECT * FROM table'\n"
+        + "    - type: ReadFromSpanner\n"
+        + "      config:\n"
+        + "        instance_id: 'my-instance-id'\n"
+        + "        database_id: 'my-database'\n"
+        + "        query: 'SELECT * FROM table'\n"
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerReadSchemaTransformProvider.java:83]: This is consistent with the other providers, so I just added it to this PR since it is the only other Java provider that did not follow this pattern. In there future, I'll make sure to separate into multiple PRs

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java (hunk 2 above)
Comment (Polber) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java:371]: The fetchSize on JdbcIO accepts Integer values, so this provider was cutting out a large range of acceptable values

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 6]
```diff
@@ -49,33 +50,174 @@ public class JdbcReadSchemaTransformProvider
...
+        + "All properties should be semi-colon-delimited (e.g. \"key1=value1;key2=value2;\")\n";
+  }
+
+  protected String baseDescription(String jdbcType) {
+    return String.format(
+        "Read from a %s source using a SQL query or by directly accessing " + "a single table.\n",
+        jdbcType);
+  }
+
+  protected String inheritedDescription(
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java:120]: The `JdbcReadSchemaTransformProvider` uses `baseDescription` but not `inheritedDescription`. `inheritedDescription` is only used by the inherited providers (MySQL, Postgres, Oracle, SQL Server), which also use the `baseDescription`.

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java [hunk 7]
```diff
@@ -49,33 +50,174 @@ public class JdbcReadSchemaTransformProvider
...
+      String prettyName, String transformName, String prefix, int port) {
```
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java:121]: Fixed

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Polber) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java:161]: I reorganized the code to avoid needing a static method, as well as remove need to override the `from()` method entirely in case there are changes to the parent method. I think it should overall be more organized now

File: sdks/python/apache_beam/yaml/generate_yaml_docs.py (hunk 4 above)
Comment (Polber) [reply sdks/python/apache_beam/yaml/generate_yaml_docs.py:207]: Added both

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java (hunk 1 above)
Comment (Polber) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcWriteSchemaTransformProvider.java:176]: Purely cosmetic

File: sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (damccorm) [reply sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcReadSchemaTransformProvider.java:120]: Right - my point is that if only one other place is ever going to use baseDescription, we could probably just inline the description in `JdbcReadSchemaTransformProvider`'s description and `inheritedDescription`. Its a small string repeat, but it saves the other providers from needing to make multiple calls (which depend on the contents) to use this.

**Review by damccorm — COMMENTED** [review]
Mostly lgtm, just one lingering comment

**Comment by damccorm** [comment]
LGTM other than checks failing

File: sdks/python/apache_beam/yaml/generate_yaml_docs.py [hunk 8]
```diff
@@ -194,12 +194,46 @@ def io_grouping_key(transform_name):
...
+  For example,
+    "Some description talking about MyTransform."
+  would be converted to
+    "Some description talking about <a href="#mytransform">MyTransform</a>"
+
+  meanwhile,
+  ```
+  - type: MyTransform
+    config:
+      ...
```
Comment (damccorm) [inline sdks/python/apache_beam/yaml/generate_yaml_docs.py:210]:
```
/runner/_work/beam/beam/sdks/python/test-suites/tox/pycommon/build/srcs/sdks/python/target/.tox-docs/docs/lib/python3.9/site-packages/apache_beam/yaml/generate_yaml_docs.py:docstring of apache_beam.yaml.generate_yaml_docs.add_transform_links:11: ERROR: Unexpected indentation.
```

from docs precommit

**Review by damccorm — COMMENTED** [review]
Looks like several precommits are still red

**Comment by Polber** [comment]
@damccorm I think failing checks are unrelated now

**Comment by damccorm** [comment]
> @damccorm I think failing checks are unrelated now

I think that's right. Rerunning to try to get a green signal before merging

**Review by damccorm — APPROVED** [review]
